onnx/shape_inference/implementation.cc

```diff
@@ -322,6 +331,13 @@ static void InferShapesImpl(
 
         // Make merged info available to further inference.
         valueTypesByName[n.output(i)] = existingType;
+
+        // For udefined output type, update both value_info and output for now
+        // If the output type is undefined, merge inferred type to it
+        iter = undefinedValueTypesByName.find(n.output(i));
+        if (iter != undefinedValueTypesByName.end()) {
+          mergeShapesAndTypes(*inferredType, iter->second);
```

Comment:
A few independent comments:
(a) Can we just do an assignment instead of calling mergeShapesAndTypes? "* iter->second = *inferredType"?
(b) Can we move this into the else-branch in line 323?

Reply:
agree with Rama on both the points...

Reply:
Both (a) and (b) are doable.
For (b), I think `inferredType` is actually determined on line 311 and this update (update original undefined output) only occurs in the else-branch on line 323. So it looks good to move there.